yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

```diff
@@ -256,6 +256,12 @@ private[spark] class ApplicationMaster(
 
   private def runDriver(securityMgr: SecurityManager): Unit = {
     addAmIpFilter()
+
+    // This must be done before SparkContext is initialized, since the CoarseGrainedSchedulerBackend
+    // is started at that time. That is what schedules the re-logins. It is scheduled only if the
+    // principal is actually setup. So we make sure it is available.
+    SparkHadoopUtil.get.setPrincipalAndKeytabForLogin(
+      System.getenv("SPARK_PRINCIPAL"), System.getenv("SPARK_KEYTAB"))
```

Comment:
Oh, but `SPARK_YARN_STAGING_DIR` is used internally, it's not user-configurable. It seems like these two options are user-configurable, right? If they're not, env variables are probably fine.